packages/fx-core/src/component/generator/utils.ts

```diff
@@ -230,8 +234,65 @@ export function getSampleRelativePath(sampleName: string): string {
   return `${sampleConfig.baseFolderName}/${sampleName}/`;
 }
 
+export function getSampleUrl(sample: SampleInfo): string {
+  if (sample.url == sampleConfig.baseUrl) {
+    return `${sample.url}${sample.id}`;
+  } else {
+    return sample.url;
+  }
+}
+
 export function zipFolder(folderPath: string): AdmZip {
   const zip = new AdmZip();
   zip.addLocalFolder(folderPath);
   return zip;
 }
+
+export async function downloadDirectory(
+  sampleUrl: string,
+  dstPath: string,
+  concurrencyLimits = sampleConcurrencyLimits,
+  retryLimits = sampleDefaultRetryLimits
+): Promise<void> {
+  //step 1: get file info
+  const urlParserRegex = /https:\/\/github.com[/]([^/]+)[/]([^/]+)[/]tree[/]([^/]+)[/](.*)/;
+  const [owner, repository, ref, sampleName] = urlParserRegex.exec(sampleUrl)!.slice(1);
+  const fileInfoUrl = `https://api.github.com/repos/${owner}/${repository}/git/trees/${ref}?recursive=1`;
+  const fileInfo = (
+    await sendRequestWithRetry(async () => {
+      return await axios.get(fileInfoUrl);
+    }, retryLimits)
+  ).data as any;
+
+  const filePrefixUrl = `https://raw.githubusercontent.com/${owner}/${repository}/${fileInfo.sha}/`;
+  const fileInfoTree = fileInfo.tree as any[];
+  const samplePaths = fileInfoTree
+    .filter((node) => node.path.startsWith(sampleName) && node.type !== "tree")
+    .map((node) => node.path);
+
+  //step 2: download files with limited concurrency
```

Comment:
IMO, it's okay to keep it in one function since these 2 steps are coupled. The step2 needs parsed results(L259) and samplePaths(269) from step1. If we want to split it into 2 functions then step2 functions will have more parameters.

Reply:
I have the some feeling that this method is too long. Let's check how to improve this code for better maintenance offline.

Reply:
This function has been splitted into 3 functions.